File: ToolKit/src/main/java/com/uday/android/toolkit/ui/PartitionListAdapter.java
package com.uday.android.toolkit.ui;
import android.widget.ArrayAdapter;
import com.uday.android.util.BlockDeviceListData;
import android.content.Context;
import java.util.ArrayList;
import android.view.View;
import android.view.ViewGroup;
import android.view.LayoutInflater;
import android.widget.TextView;
import com.uday.android.toolkit.R;
import android.widget.ImageView;
import android.view.animation.Animation;
import com.uday.android.util.OptAnimationLoader;

public class PartitionListAdapter extends ArrayAdapter<BlockDeviceListData>
{
	private ArrayList<BlockDeviceListData> partitionList;
	private Context context;
	private int layout_res;
	
	public PartitionListAdapter(Context context,int layoutRes,ArrayList<BlockDeviceListData> partitionListData){
		super(context,layoutRes,partitionListData);
		this.partitionList=partitionListData;
		this.context=context;
		this.layout_res=layoutRes;
	}
	
	@Override
	public View getView(final int position, View convertView, ViewGroup parent)
	{
		
		LayoutInflater inflater=(LayoutInflater)getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
		final View row=inflater.inflate(layout_res,parent,false);
		final BlockDeviceListData blockDevice=partitionList.get(position);
		
		((TextView)row.findViewById(R.id.part_name)).setText(blockDevice.getName());
		((TextView)row.findViewById(R.id.block_dev)).setText(blockDevice.getBlock().getAbsolutePath());
		((TextView)row.findViewById(R.id.part_size)).setText(blockDevice.getSizeStr());
		((TextView)row.findViewById(R.id.start_addr)).setText(blockDevice.getStartStr());
		((TextView)row.findViewById(R.id.end_addr)).setText(blockDevice.getEndStr());
		row.findViewById(R.id.size_expand).setOnClickListener(new View.OnClickListener(){
			@Override
			public void onClick(View p1){
				final View addrs=row.findViewById(R.id.part_size_addr);
				if(addrs.getVisibility()==View.GONE){
					Animation animation=OptAnimationLoader.loadAnimation(getContext(),R.anim.expand_to_down);
					addrs.setVisibility(View.VISIBLE);
					addrs.startAnimation(animation);
				}else{
					Animation animation=OptAnimationLoader.loadAnimation(getContext(),R.anim.shrink_to_top);
					animation.setAnimationListener(new Animation.AnimationListener(){
							@Override
							public void onAnimationStart(Animation p1){}
							public void onAnimationEnd(Animation p1){
								addrs.setVisibility(View.GONE);
							}
							public void onAnimationRepeat(Animation p1){}
					});
					addrs.startAnimation(animation);
				}
			}
		});
		return row;
	}
}

File: app/src/main/java/com/uday/android/toolkit/ui/PartitionListAdapter.kt
package com.uday.android.toolkit.ui

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.ArrayAdapter
import android.widget.HorizontalScrollView
import android.widget.ImageView
import android.widget.TextView
import com.uday.android.toolkit.R
import com.uday.android.util.BlockDeviceListData
import java.util.*

class PartitionListAdapter(
    private val contex: Context,
    private val layout_res: Int,
    private val partitionList: ArrayList<BlockDeviceListData>) : ArrayAdapter<BlockDeviceListData>(contex, layout_res, partitionList) {

    @SuppressLint("ViewHolder")
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {

        val inflater =
            context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val row = inflater.inflate(layout_res, parent, false)
        val blockDevice = partitionList[position]

        (row.findViewById(R.id.part_name) as TextView).text = blockDevice.getName()
        (row.findViewById(R.id.block_dev) as TextView).text = blockDevice.getBlock()!!.absolutePath
        (row.findViewById(R.id.part_size) as TextView).text = blockDevice.sizeStr
        (row.findViewById(R.id.start_addr) as TextView).text = blockDevice.startStr
        (row.findViewById(R.id.end_addr) as TextView).text = blockDevice.endStr
        row.findViewById<ImageView>(R.id.size_expand).setOnClickListener {
            val addrs = row.findViewById<HorizontalScrollView>(R.id.part_size_addr)
            if (addrs.visibility == View.GONE) {
                val animation =
                    AnimationUtils.loadAnimation(context, R.anim.expand_to_down)
                addrs.visibility = View.VISIBLE
                addrs.startAnimation(animation)
            } else {
                val animation = AnimationUtils.loadAnimation(context, R.anim.shrink_to_top)
                animation!!.setAnimationListener(object : Animation.AnimationListener {
                    override fun onAnimationStart(p1: Animation) {}
                    override fun onAnimationEnd(p1: Animation) {
                        addrs.visibility = View.GONE
                    }

                    override fun onAnimationRepeat(p1: Animation) {}
                })
                addrs.startAnimation(animation)
            }
        }
        return row
    }
}
